src/psyclone/dynamo0p3.py

```diff
@@ -3910,22 +3935,37 @@ def _stub_declarations(self, parent):
                                intent="in",
                                dimension=",".join(basis_arrays[basis]),
                                entity_decls=[basis]))
-        for qr_shape in VALID_QUADRATURE_SHAPES:
-            if qr_shape.lower() == "gh_quadrature_xyoz":
-                if qr_shape not in self._qr_vars:
-                    continue
+
+        for shape in self._qr_vars:
+            qr_name = "_qr_" + shape.split("_")[-1]
+            if shape == "gh_quadrature_xyoz":
+                parent.add(DeclGen(parent, datatype="real",
+                                   kind=api_config.default_kind["real"],
+                                   intent="in", dimension="np_xy"+qr_name,
+                                   entity_decls=["weights_xy"+qr_name]))
+                parent.add(DeclGen(parent, datatype="real",
+                                   kind=api_config.default_kind["real"],
+                                   intent="in", dimension="np_z"+qr_name,
+                                   entity_decls=["weights_z"+qr_name]))
+            elif shape == "gh_quadrature_face":
                 parent.add(DeclGen(parent, datatype="real",
                                    kind=api_config.default_kind["real"],
-                                   intent="in", dimension="np_xy",
-                                   entity_decls=["weights_xy"]))
+                                   intent="in",
+                                   dimension=",".join(["np_xyz"+qr_name,
+                                                       "nfaces"+qr_name]),
+                                   entity_decls=["weights_xyz"+qr_name]))
+            elif shape == "gh_quadrature_edge":
                 parent.add(DeclGen(parent, datatype="real",
                                    kind=api_config.default_kind["real"],
-                                   intent="in", dimension="np_z",
-                                   entity_decls=["weights_z"]))
+                                   intent="in",
+                                   dimension=",".join(["np_xyz"+qr_name,
+                                                       "nedges"+qr_name]),
+                                   entity_decls=["weights_xyz"+qr_name]))
             else:
                 raise GenerationError(
-                    "Quadrature shapes other than GH_QUADRATURE_XYoZ are not "
-                    "yet supported - got '{0}'".format(qr_shape))
+                    "Quadrature shapes other than GH_QUADRATURE_XYoZ, "
```

Comment:
Good idea - an InternalError is more applicable (since this error condition is a defensive measure and shouldn't be triggered by user error) so I've updated the code and I've added a list of the supported shapes. The error condition itself is covered by the test suite.